File: src/go_lcm/lcm.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/signal"
	"strings"
	"syscall"

	"github.com/confluentinc/confluent-kafka-go/kafka"
	_ "github.com/lib/pq"
)

// InputMessage represents the JSON message structure for input.
type InputMessage struct {
	Numbers []int `json:"numbers"`
	Sum     int   `json:"sum"`
	Product int   `json:"product"`
}

// OutputMessage represents the JSON message structure for output.
type OutputMessage struct {
	Numbers []int `json:"numbers"`
	Sum     int   `json:"sum"`
	Product int   `json:"product"`
	LCM     int   `json:"lcm"`
}

func main() {
	// Check if required environment variables are set
	requiredEnvVars := []string{"DB_USER", "DB_PASSWORD"}
	missingVars := []string{}
	for _, envVar := range requiredEnvVars {
		if os.Getenv(envVar) == "" {
			missingVars = append(missingVars, envVar)
		}
	}

	if len(missingVars) > 0 {
		log.Printf("Missing required environment variables: %s", strings.Join(missingVars, ", "))
		os.Exit(1)
	}

	bootstrapServers := getEnv("KAFKA_BOOTSTRAP_SERVERS", "localhost:9002")
	inputTopic := getEnv("KAFKA_TOPIC_INPUT", "stage3")
	outputTopic := getEnv("KAFKA_TOPIC_OUTPUT", "stage4")
	dbHost := getEnv("DB_HOST", "localhost")
	dbPort := getEnv("DB_PORT", "5432")
	dbName := getEnv("DB_NAME", "stage")
	dbUser := os.Getenv("DB_USER")
	dbPassword := os.Getenv("DB_PASSWORD")

	consumer, err := kafka.NewConsumer(&kafka.ConfigMap{
		"bootstrap.servers": bootstrapServers,
		"group.id":          "stage3",
		"auto.offset.reset": "earliest",
	})
	if err != nil {
		log.Fatalf("Failed to create consumer: %v", err)
	}
	defer consumer.Close()

	producer, err := kafka.NewProducer(&kafka.ConfigMap{
		"bootstrap.servers": bootstrapServers,
	})
	if err != nil {
		log.Fatalf("Failed to create producer: %v", err)
	}
	defer producer.Close()

	// Connect to the Postgres database
	dbInfo := fmt.Sprintf("host=%s port=%s dbname=%s user=%s password=%s sslmode=disable", dbHost, dbPort, dbName, dbUser, dbPassword)
	db, err := sql.Open("postgres", dbInfo)
	if err != nil {
		log.Fatalf("Failed to connect to the database: %v", err)
	}
	defer db.Close()

	//Ensure the table exists
	err = createTableIfNotExists(db)
	if err != nil {
		log.Fatalf("Failed to create table: %v", err)
	}

	// Subscribe to the input topic
	err = consumer.SubscribeTopics([]string{inputTopic}, nil)
	if err != nil {
		log.Fatalf("Failed to subscribe to input topic: %v", err)
	}

	// Handle termination signals to gracefully exit
	signals := make(chan os.Signal, 1)
	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)

	// Start consuming messages
	for {
		select {
		case sig := <-signals:
			log.Printf("Termination signal received: %v", sig)
			return
		default:
			msg, err := consumer.ReadMessage(-1)
			if err != nil {
				log.Printf("Consumer error: %v", err)
				continue
			}

			var inputMsg InputMessage
			err = json.Unmarshal(msg.Value, &inputMsg)
			if err != nil {
				log.Printf("Failed to parse JSON message: %v", err)
				continue
			}

			result := calculateLCM(inputMsg.Numbers)
			outputMsg := OutputMessage{
				Numbers: inputMsg.Numbers,
				LCM:     result,
			}

			outputJSON, err := json.Marshal(outputMsg)
			if err != nil {
				log.Printf("Failed to encode JSON message: %v", err)
				continue
			}

			// Produce the result to the output topic
			deliveryChan := make(chan kafka.Event, 10000)
			err = producer.Produce(&kafka.Message{
				TopicPartition: kafka.TopicPartition{Topic: &outputTopic, Partition: kafka.PartitionAny},
				Value:          outputJSON},
				deliveryChan)
			if err != nil {
				log.Printf("Failed to produce message: %v", err)
				continue
			}

			// Wait for the delivery report
			e := <-deliveryChan
			if m, ok := e.(*kafka.Message); ok {
				if m.TopicPartition.Error != nil {
					log.Printf("Failed to deliver message: %v", m.TopicPartition.Error)
				} else {
					log.Printf("Message delivered to topic %s [%d] at offset %v\n",
						*m.TopicPartition.Topic, m.TopicPartition.Partition, m.TopicPartition.Offset)
				}
			}

			// Save the input numbers and LCM to the Postgres table
			err = saveToPostgres(db, inputMsg.Numbers, result)
			if err != nil {
				log.Printf("Failed to save to Postgres: %v", err)
				continue
			}
		}
	}
}

func calculateLCM(numbers []int) int {
	lcm := numbers[0]
	for i := 1; i < len(numbers); i++ {
		lcm = calculateTwoLCM(lcm, numbers[i])
	}
	return lcm
}

func calculateTwoLCM(a, b int) int {
	return a * b / calculateGCD(a, b)
}

func calculateGCD(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func createTableIfNotExists(db *sql.DB) error {
	query := `
		CREATE TABLE IF NOT EXISTS lcm_results (
			numbers TEXT,
			stage TEXT,
			result TEXT
		);`
	_, err := db.Exec(query)
	return err
}

func saveToPostgres(db *sql.DB, numbers []int, result int) error {
	numbersJSON, err := json.Marshal(numbers)
	if err != nil {
		return fmt.Errorf("failed to marshal numbers to JSON: %v", err)
	}

	query := `
		INSERT INTO results (numbers, stage, result)
		VALUES ($1, 'stage3', $2);`
	_, err = db.Exec(query, numbersJSON, result)
	if err != nil {
		return fmt.Errorf("failed to save to Postgres: %v", err)
	}

	return nil
}

func getEnv(key, fallback string) string {
	if value, ok := os.LookupEnv(key); ok {
		return value
	}
	return fallback
}

File: src/lcm/lcm.py
import os
import json
import psycopg2
from confluent_kafka import Consumer, Producer

class LCMMicroservice:
    def __init__(self):

        # Check if required environment variables are set
        required_env_vars = ['STAGE_NUMBER', 'DB_NAME', 'DB_USER', 'DB_PASSWORD', 'INPUT_TOPIC', 'OUTPUT_TOPIC']
        missing_vars = [var for var in required_env_vars if not os.environ.get(var)]
        if missing_vars:
            print(f"Missing required environment variables: {', '.join(missing_vars)}")
            exit(1)

        self.stage = os.environ.get('STAGE_NUMBER')
        self.bootstrap_servers = os.environ.get('BOOTSTRAP_SERVERS', 'localhost:9092')
        self.input_topic = os.environ.get('INPUT_TOPIC')
        self.output_topic = os.environ.get('OUTPUT_TOPIC')
        self.db_host = os.environ.get('DB_HOST', 'localhost')
        self.db_port = os.environ.get('DB_PORT', '5432')
        self.db_name = os.environ.get('DB_NAME')
        self.db_user = os.environ.get('DB_USER')
        self.db_password = os.environ.get('DB_PASSWORD')
        self.consumer = None
        self.producer = None


    def start(self):
        # Create a Kafka consumer
        self.consumer = Consumer({
            'bootstrap.servers': self.bootstrap_servers,
            'group.id': 'lcm_microservice'
        })
        self.consumer.subscribe([self.input_topic])

        # Create a Kafka producer
        self.producer = Producer({'bootstrap.servers': self.bootstrap_servers})

        # Start consuming messages
        while True:
            message = self.consumer.poll(1.0)

            if message is None:
                continue

            if message.error():
                print(f"Consumer error: {message.error()}")
                continue

            print("Got message from Kafka: %s" % (message.value()))
            # Process the received message
            data = self.deserialize_message(message.value())

            data['lcm'] = self.calculate_lcm(data['numbers'])

            # Save the result to the Postgres table
            self.insert_into_postgres(data['numbers'], data['lcm'])

            # Publish the updated JSON object to the output topic
            self.publish_result(data)

    def deserialize_message(self, message):
        # Deserialize JSON message
        return json.loads(message)

    def insert_into_postgres(self, numbers, data):
        # Connect to the Postgres database
        conn = psycopg2.connect(
            host=self.db_host,
            port=self.db_port,
            dbname=self.db_name,
            user=self.db_user,
            password=self.db_password
        )
        cursor = conn.cursor()
        create_table_query = """
            CREATE TABLE IF NOT EXISTS results (
                numbers TEXT,
                stage TEXT,
                result TEXT
            )
        """
        cursor.execute(create_table_query)

        # Insert the result into the table
        cursor.execute(f"INSERT INTO results (numbers, stage, result) VALUES (%s, %s, %s)", (json.dumps(numbers), self.stage, data))
        print("SQL Query: " + str(cursor.query))

        # Commit the transaction and close the connection
        conn.commit()
        cursor.close()
        conn.close()

    def publish_result(self, data):
        # Serialize JSON object
        payload = json.dumps(data)

        # Publish the result to the output topic
        self.producer.produce(self.output_topic, value=payload.encode('utf-8'))
        print("Produce message on topic %s" % (self.output_topic))

        # Flush the producer tomake sure the message is sent
        self.producer.flush()

    def calculate_lcm(self, numbers):
        def gcd(a, b):
            while b:
                a, b = b, a % b
            return a

        lcm = numbers[0]
        for num in numbers[1:]:
            lcm = lcm * num // gcd(lcm, num)
        return lcm


if __name__ == '__main__':
    microservice = LCMMicroservice()

    microservice.start()